File: 总题库/79.WordSearch.py
class Solution:
    def __init__(self):
        super().__init__()
        self.addX = [1, -1, 0, 0]
        self.addY = [0, 0, 1, -1]
    
    def dfs(self, board: List[List[str]], word: str, index: int, x: int, y: int) -> bool:
        if index == len(word):
            return True;

        rows = len(board)
        cols = len(board[0])

        for i in range(4):
            newX = x + self.addX[i]
            newY = y + self.addY[i]
            if newX >= 0 and newX < rows and newY >= 0 and newY < cols:
                if board[newX][newY] == word[index]:
                    temp = board[newX][newY]
                    board[newX][newY] = ' '
                    if (self.dfs(board, word, index + 1, newX, newY)):
                        return True
                    board[newX][newY] = temp
        return False

    def exist(self, board: List[List[str]], word: str) -> bool:
        rows = len(board)
        cols = len(board[0])
        for i in range(rows):
            for j in range(cols):
                if board[i][j] == word[0]:
                    temp = board[i][j]
                    board[i][j] = ' '
                    if (self.dfs(board, word, 1, i, j)):
                        return True
                    board[i][j] = temp
        return False

File: 总题库/79.WordSearch.cpp
class Solution {
    int addX[4] = {1, -1, 0, 0};
    int addY[4] = {0, 0, 1, -1};

    bool dfs(vector<vector<char>>& board, string word, int index, int x, int y) {
        if (index == word.length()) return true;

        int rows = board.size(), cols = board[0].size();
        for (int i = 0; i < 4; i++) {
            int newX = x + addX[i], newY = y + addY[i];
            if (newX >= 0 && newX < rows && newY >= 0 && newY < cols) {
                if (board[newX][newY] == word[index]) {
                    char temp = board[newX][newY];
                    board[newX][newY] = ' ';
                    if (dfs(board, word, index + 1, newX, newY)) return true;
                    board[newX][newY] = temp;
                }
            }
        }
        return false;
    }

public:
    bool exist(vector<vector<char>>& board, string word) {
        int rows = board.size(), cols = board[0].size();
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (board[i][j] == word[0]) {
                    char temp = board[i][j];
                    board[i][j] = ' ';
                    if (dfs(board, word, 1, i, j)) return true;
                    board[i][j] = temp;
                }
            }
        }
        return false;
    }
};